`conan install` corrupts the cache and prevents other projects from compiling

### Environment details

* Operating System+version: Windows 11 22H2
* Compiler+version: Visual Studio 2022
* Conan version: 2.0.13
* Python version: 3.11.2


### Steps to reproduce

We are migrating from Conan v1 to v2 and encountered a bug. Here is a simple demonstration: [ConanCache.zip](https://github.com/conan-io/conan/files/12904230/ConanCache.zip).

conanfile.txt:
```
[requires]
fmt/10.1.1

[generators]
MSBuildDeps

[options]
fmt/*:header_only=True
```
First, we compile the dependencies for Project1, and Project1 works fine. But after we compile the dependencies for Project2, Project1 fails to compile because it cannot find <fmt/format.h>.

This error only occurs under specific configurations. If we remove the `header_only` option or use `--build=missing` in `conan install`, everything works fine.

### Logs

_No response_

Hello, thank you for your response! Please run build.bat in the "Visual Studio 2022 Developer Command Prompt", or you can temporarily add msbuild to the path.

> As a side note, I wouldn't recommend to use --build=*, because it unnecessarily re-build dependencies from source again and again, taking time and storage.

We are using Directory.Build.props to customize the behavior of msbuild, so we have to compile all the dependencies ourselves.

Here is the output:

```
C:\Users\Xu\Downloads\ConanCache>conan install conanfile.txt --output-folder Project1\.conan --build=* -s build_type=Release -s arch=x86_64 --update

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows


======== Computing dependency graph ========
fmt/10.1.1: Checking remote: conancenter
Graph root
    conanfile.txt: C:\Users\Xu\Downloads\ConanCache\conanfile.txt
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243 - Cache (conancenter)

======== Computing necessary packages ========
fmt/10.1.1: Forced build from source
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709 - Build

======== Installing packages ========

-------- Installing package fmt/10.1.1 (1 of 1) --------
fmt/10.1.1: Building from source
fmt/10.1.1: Package fmt/10.1.1:da39a3ee5e6b4b0d3255bfef95601890afd80709
fmt/10.1.1: Copying sources to build folder
fmt/10.1.1: Building your package in C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\b
fmt/10.1.1: Calling generate()
fmt/10.1.1: Generators folder: C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\b\build-release\conan
fmt/10.1.1: Generating aggregated env files
fmt/10.1.1: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
fmt/10.1.1: Calling build()
fmt/10.1.1: apply_conandata_patches(): No patches defined in conandata
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' built
fmt/10.1.1: Build folder C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\b\build-release
fmt/10.1.1: Generating the package
fmt/10.1.1: Packaging in folder C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\p
fmt/10.1.1: Calling package()
fmt/10.1.1: package(): Packaged 13 '.h' files
fmt/10.1.1: package(): Packaged 1 '.rst' file: LICENSE.rst
fmt/10.1.1: Created package revision 6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' created
fmt/10.1.1: Full package reference: fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709#6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package folder C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\p
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: fmt/10.1.1

======== Finalizing install (deploy, generators) ========
conanfile.txt: Writing generators to C:\Users\Xu\Downloads\ConanCache\Project1\.conan
conanfile.txt: Generator 'MSBuildDeps' calling 'generate()'
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
Install finished successfully

C:\Users\Xu\Downloads\ConanCache>REM WORKS

C:\Users\Xu\Downloads\ConanCache>msbuild -p:Configuration=Release;Platform=x64 Project1\Project1.sln
MSBuild version 17.7.2+d6990bcfa for .NET Framework
Build started 2023/10/15 9:20:08.

Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" on node 1 (default targets).
ValidateSolutionConfiguration:
  Building solution configuration "Release|x64".
Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (1) is building "C:\Users\Xu\Downloads\ConanCache\Proj
ect1\Project1.vcxproj" (2) on node 1 (default targets).
InitializeBuildStatus:
  Touching "x64\Release\Project1.tlog\unsuccessfulbuild".
ClCompile:
  C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64\CL.exe /c /IC:/User
  s/Xu/.conan2/p/b/fmtacd454cb5451f/p/include /Zi /nologo /W3 /WX- /diagnostics:column /sdl /O2 /Oi /GL /D FMT_HEADER_O
  NLY=1 /D NDEBUG /D _CONSOLE /D _UNICODE /D UNICODE /Gm- /EHsc /MD /GS /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:in
  line /permissive- /Fo"x64\Release\\" /Fd"x64\Release\vc143.pdb" /external:W3 /Gd /TP /FC /errorReport:queue  Project1
  .cpp
  Project1.cpp
Link:
  C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64\link.exe /ERRORREPO
  RT:QUEUE /OUT:"C:\Users\Xu\Downloads\ConanCache\Project1\x64\Release\Project1.exe" /NOLOGO kernel32.lib user32.lib gd
  i32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib /M
  ANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"C:\Users\Xu\Downloads\ConanCac
  he\Project1\x64\Release\Project1.pdb" /SUBSYSTEM:CONSOLE /OPT:REF /OPT:ICF /LTCG:incremental /LTCGOUT:"x64\Release\Pr
  oject1.iobj" /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:\Users\Xu\Downloads\ConanCache\Project1\x64\Release\Project1.
  lib" /MACHINE:X64  x64\Release\Project1.obj
  ??????
  853 of 1121 functions (76.1%) were compiled, the rest were copied from previous compilation.
    0 functions were new in current compilation
    0 functions had inline decision re-evaluated but remain unchanged
  ????????
  Project1.vcxproj -> C:\Users\Xu\Downloads\ConanCache\Project1\x64\Release\Project1.exe
FinalizeBuildStatus:
  Deleting file "x64\Release\Project1.tlog\unsuccessfulbuild".
  Touching "x64\Release\Project1.tlog\Project1.lastbuildstate".
Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj" (default targets).

Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.79

C:\Users\Xu\Downloads\ConanCache>conan install conanfile.txt --output-folder Project2\.conan --build=* -s build_type=Release -s arch=x86_64 --update

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows


======== Computing dependency graph ========
fmt/10.1.1: Checking remote: conancenter
Graph root
    conanfile.txt: C:\Users\Xu\Downloads\ConanCache\conanfile.txt
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243 - Cache (conancenter)

======== Computing necessary packages ========
fmt/10.1.1: Forced build from source
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709 - Build

======== Installing packages ========

-------- Installing package fmt/10.1.1 (1 of 1) --------
fmt/10.1.1: Building from source
fmt/10.1.1: Package fmt/10.1.1:da39a3ee5e6b4b0d3255bfef95601890afd80709
fmt/10.1.1: Copying sources to build folder
fmt/10.1.1: Building your package in C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\b
fmt/10.1.1: Calling generate()
fmt/10.1.1: Generators folder: C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\b\build-release\conan
fmt/10.1.1: Generating aggregated env files
fmt/10.1.1: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
fmt/10.1.1: Calling build()
fmt/10.1.1: apply_conandata_patches(): No patches defined in conandata
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' built
fmt/10.1.1: Build folder C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\b\build-release
fmt/10.1.1: Generating the package
fmt/10.1.1: Packaging in folder C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\p
fmt/10.1.1: Calling package()
fmt/10.1.1: package(): Packaged 13 '.h' files
fmt/10.1.1: package(): Packaged 1 '.rst' file: LICENSE.rst
fmt/10.1.1: Created package revision 6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' created
fmt/10.1.1: Full package reference: fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709#6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package folder C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\p
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: fmt/10.1.1

======== Finalizing install (deploy, generators) ========
conanfile.txt: Writing generators to C:\Users\Xu\Downloads\ConanCache\Project2\.conan
conanfile.txt: Generator 'MSBuildDeps' calling 'generate()'
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
Install finished successfully

C:\Users\Xu\Downloads\ConanCache>REM FAIL!!!

C:\Users\Xu\Downloads\ConanCache>msbuild -p:Configuration=Release;Platform=x64 Project1\Project1.sln
MSBuild version 17.7.2+d6990bcfa for .NET Framework
Build started 2023/10/15 9:20:11.

Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" on node 1 (default targets).
ValidateSolutionConfiguration:
  Building solution configuration "Release|x64".
Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (1) is building "C:\Users\Xu\Downloads\ConanCache\Proj
ect1\Project1.vcxproj" (2) on node 1 (default targets).
InitializeBuildStatus:
  Creating "x64\Release\Project1.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
  Touching "x64\Release\Project1.tlog\unsuccessfulbuild".
ClCompile:
  C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64\CL.exe /c /IC:/User
  s/Xu/.conan2/p/b/fmtacd454cb5451f/p/include /Zi /nologo /W3 /WX- /diagnostics:column /sdl /O2 /Oi /GL /D FMT_HEADER_O
  NLY=1 /D NDEBUG /D _CONSOLE /D _UNICODE /D UNICODE /Gm- /EHsc /MD /GS /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:in
  line /permissive- /Fo"x64\Release\\" /Fd"x64\Release\vc143.pdb" /external:W3 /Gd /TP /FC /errorReport:queue  Project1
  .cpp
  Project1.cpp
C:\Users\Xu\Downloads\ConanCache\Project1\Project1.cpp(2,10): fatal  error C1083: ????????: "fmt/format.h": No such fil
e or directory [C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj]
Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj" (default targets) -- FAILED.

Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (default targets) -- FAILED.


Build FAILED.

"C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (default target) (1) ->
"C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj" (default target) (2) ->
(ClCompile target) ->
  C:\Users\Xu\Downloads\ConanCache\Project1\Project1.cpp(2,10): fatal  error C1083: ????????: "fmt/format.h": No such f
ile or directory [C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.51
```

--- a/conan/internal/cache/cache.py
+++ b/conan/internal/cache/cache.py
@@ -1,5 +1,6 @@
 import hashlib
 import os
+import shutil
 import uuid
 
 from conan.internal.cache.conan_reference_layout import RecipeLayout, PackageLayout
@@ -185,9 +186,16 @@ class DataCache:
             self._db.create_package(relpath, pref, build_id)
         except ConanReferenceAlreadyExistsInDB:
             # TODO: Optimize this into 1 single UPSERT operation
-            # This was exported before, making it latest again, update timestamp
+            # There was a previous package folder for this same package reference (and prev)
             pkg_layout = self.get_package_layout(pref)
+            # We remove the old one and move the new one to the path of the previous one
+            # this can be necessary in case of new metadata or build-folder because of "build_id()"
             pkg_layout.remove()
+            shutil.move(layout.base_folder, pkg_layout.base_folder)  # clean unused temporary build
+            layout._base_folder = pkg_layout.base_folder  # reuse existing one
+            # TODO: The relpath would be the same as the previous one, it shouldn't be ncessary to
+            #  update it, the update_package_timestamp() can be simplified and path dropped
+            relpath = os.path.relpath(layout.base_folder, self._base_folder)
             self._db.update_package_timestamp(pref, path=relpath, build_id=build_id)
 
     def assign_rrev(self, layout: RecipeLayout):

--- a/conans/test/integration/cache/test_same_pref_removal.py
+++ b/conans/test/integration/cache/test_same_pref_removal.py
@@ -1,0 +1,21 @@
+import os
+
+from conans.test.assets.genconanfile import GenConanfile
+from conans.test.utils.tools import TestClient
+
+
+def test_same_pref_removal():
+    c = TestClient()
+    c.save({"conanfile.py": GenConanfile("pkg", "0.1")})
+    c.run("export .")
+    c.run("install --requires=pkg/0.1 --build=pkg*")
+    layout = c.created_layout()
+    pkg_folder1 = layout.package()
+    assert os.path.exists(os.path.join(pkg_folder1, "conanmanifest.txt"))
+    assert os.path.exists(os.path.join(pkg_folder1, "conaninfo.txt"))
+    c.run("install --requires=pkg/0.1 --build=pkg*")
+    layout = c.created_layout()
+    pkg_folder2 = layout.package()
+    assert pkg_folder1 == pkg_folder2
+    assert os.path.exists(os.path.join(pkg_folder1, "conanmanifest.txt"))
+    assert os.path.exists(os.path.join(pkg_folder1, "conaninfo.txt"))

--- a/conans/test/integration/command_v2/test_cache_clean.py
+++ b/conans/test/integration/command_v2/test_cache_clean.py
@@ -60,8 +60,8 @@ def test_cache_clean_all():
 
 def test_cache_multiple_builds_same_prev_clean():
     """
-    Different consecutive builds will create different folders, even if for the
-    same exact prev, leaving trailing non-referenced folders
+    Different consecutive builds will create exactly the same folder, for the
+    same exact prev, not leaving trailing non-referenced folders
     """
     c = TestClient()
     c.save({"conanfile.py": GenConanfile("pkg", "0.1")})
@@ -75,7 +75,7 @@ def test_cache_multiple_builds_same_prev_clean():
     c.run("cache path pkg/0.1:da39a3ee5e6b4b0d3255bfef95601890afd80709")
     path2 = str(c.stdout)
     assert path2 in create_out
-    assert path1 != path2
+    assert path1 == path2
 
     builds_folder = os.path.join(c.cache_folder, "p", "b")
     assert len(os.listdir(builds_folder)) == 1  # only one build

